File: android/app/src/main/java/info/bagen/dwebbrowser/microService/browser/mwebview/MultiWebViewNMM.kt
package info.bagen.dwebbrowser.microService.browser.mwebview

import info.bagen.dwebbrowser.microService.core.AndroidNativeMicroModule
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus
import org.dweb_browser.browserUI.download.DownLoadObserver
import org.dweb_browser.dwebview.base.ViewItem
import org.dweb_browser.dwebview.serviceWorker.emitEvent
import org.dweb_browser.helper.ComparableWrapper
import org.dweb_browser.helper.ImageResourcePurposes
import org.dweb_browser.helper.StrictImageResource
import org.dweb_browser.helper.enumToComparable
import org.dweb_browser.helper.ioAsyncExceptionHandler
import org.dweb_browser.helper.printDebug
import org.dweb_browser.microservice.core.BootstrapContext
import org.dweb_browser.microservice.core.MicroModule
import org.dweb_browser.microservice.help.types.MICRO_MODULE_CATEGORY
import org.dweb_browser.microservice.help.types.MMID
import org.dweb_browser.microservice.ipc.Ipc
import org.dweb_browser.window.core.WindowState
import org.dweb_browser.window.core.constant.UUID
import org.dweb_browser.window.core.constant.WindowConstants
import org.dweb_browser.window.core.createWindowAdapterManager
import org.dweb_browser.window.render.emitFocusOrBlur
import org.http4k.core.Method
import org.http4k.core.Response
import org.http4k.core.Status
import org.http4k.lens.Query
import org.http4k.lens.string
import org.http4k.routing.bind
import org.http4k.routing.routes

fun debugMultiWebView(tag: String, msg: Any? = "", err: Throwable? = null) =
  printDebug("mwebview", tag, msg, err)

class MultiWebViewNMM :
  AndroidNativeMicroModule("mwebview.browser.dweb", "Multi Webview Renderer") {
  init {
    short_name = "MWebview"
    categories =
      mutableListOf(MICRO_MODULE_CATEGORY.Service, MICRO_MODULE_CATEGORY.Render_Service)
  }
  private val ioAsyncScope = MainScope() + ioAsyncExceptionHandler

  companion object {
    private val controllerMap = mutableMapOf<MMID, MultiWebViewController>()

    /**获取当前的controller, 只能给nativeUI 使用，因为他们是和mwebview绑定在一起的
     */
    fun getCurrentWebViewController(mmid: MMID): MultiWebViewController? {
      return controllerMap[mmid]
    }
  }

  override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {
    // 打开webview
    val queryUrl = Query.string().required("url")
    val queryWebviewId = Query.string().required("webview_id")

    apiRouting = routes(
      // 打开一个 webview，并将它以 窗口window 的标准进行展示
      "/open" bind Method.GET to defineHandler { request, ipc ->
        val url = queryUrl(request)
        val remoteMm = ipc.asRemoteInstance()
          ?: throw Exception("mwebview.browser.dweb/open should be call by locale")
        ipc.onClose {
          debugMultiWebView("/open", "listen ipc close destroy window")
          val controller = controllerMap[ipc.remote.mmid]
          controller?.destroyWebView()
        }

        val (viewItem, controller) = openDwebView(url, remoteMm, ipc)
        return@defineHandler ViewItemResponse(viewItem.webviewId, controller.win.id)
      },
      // 关闭指定 webview 窗口
      "/close" bind Method.GET to defineHandler { request, ipc ->
        val webviewId = queryWebviewId(request)
        val remoteMmid = ipc.remote.mmid
        debugMultiWebView("/close", "webviewId:$webviewId,mmid:$remoteMmid")
        closeDwebView(remoteMmid, webviewId)
      },
      "/close/app" bind Method.GET to defineHandler { request, ipc ->
        val controller = controllerMap[ipc.remote.mmid] ?: return@defineHandler false;
        controller.destroyWebView()
      },
      // 界面没有关闭，用于重新唤醒
      "/activate" bind Method.GET to defineHandler { request, ipc ->
        val remoteMmid = ipc.remote.mmid
        val controller = controllerMap[remoteMmid] ?: return@defineHandler false;
        debugMultiWebView("/activate", "激活 ${controller.ipc.remote.mmid}")
        // TODO 将当前的界面移动到最上层
        //  controller.ipc.postMessage(IpcEvent.fromUtf8(EIpcEvent.Activity.event, ""))
        controller.win.emitFocusOrBlur(true)

        return@defineHandler Response(Status.OK)
      },
    )
  }

  data class ViewItemResponse(val webviewId: String, val wid: UUID)

  override suspend fun _shutdown() {
    apiRouting = null
    ioAsyncScope.cancel()
  }

  private suspend fun openDwebView(
    url: String,
    remoteMm: MicroModule,
    ipc: Ipc,
  ): Pair<ViewItem, MultiWebViewController> {
    val remoteMmid = remoteMm.mmid
    debugMultiWebView("/open", "remote-mmid: $remoteMmid / url:$url")

    val controller = controllerMap.getOrPut(remoteMmid) {
      val win = createWindowAdapterManager.createWindow(
        WindowState(
          WindowConstants(
            owner = ipc.remote.mmid,
            ownerVersion = ipc.remote.version,
            provider = mmid,
            microModule = this
          )
        ).also { state ->
          /**
           * 挑选合适的图标作为应用的图标
           */
          val iconResource = ipc.remote.icons?.let { icons ->
            val comparableBuilder =
              ComparableWrapper.Builder<StrictImageResource> { imageResource ->
                mapOf(
                  "purpose" to enumToComparable(
                    imageResource.purpose,
                    listOf(
                      ImageResourcePurposes.Maskable,
                      ImageResourcePurposes.Any,
                      ImageResourcePurposes.Monochrome
                    )
                  ).first(),
                  "type" to enumToComparable(
                    imageResource.type,
                    listOf("image/svg+xml", "image/png", "image/jpeg", "image/*")
                  ),
                  "area" to imageResource.sizes.last().let {
                    -it.width * it.height
                  }
                )
              }

            icons.minOfOrNull { comparableBuilder.build(StrictImageResource.from(it)) }?.value
          }
          if (iconResource != null) {
            state.iconUrl = iconResource.src
            state.iconMaskable = iconResource.purpose.contains(ImageResourcePurposes.Maskable)
            state.iconMonochrome = iconResource.purpose.contains(ImageResourcePurposes.Monochrome)
          } else {

          }
        }
      )

      MultiWebViewController(win, ipc, remoteMm, this).also { controller ->
        /// 窗口销毁的时候，释放这个Controller
        win.onClose {
          controllerMap.remove(remoteMmid)
        }
        ioAsyncScope.launch {
          controller.downLoadObserver = DownLoadObserver(remoteMmid).apply {
            observe { listener ->
              controller.lastViewOrNull?.webView?.let { dWebView ->
                emitEvent(
                  dWebView, listener.downLoadStatus.toServiceWorkerEvent(), listener.progress
                )
              }
            }
          }
        }
      }
    }

    return Pair(controller.openWebView(url), controller)
  }

  private suspend fun closeDwebView(remoteMmid: String, webviewId: String): Boolean {
    return controllerMap[remoteMmid]?.closeWebView(webviewId) ?: false
  }
}

File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/Mwebview/MultiWebViewNMM.cs
using System.Net;
using DwebBrowser.MicroService.Http;

#nullable enable

namespace DwebBrowser.MicroService.Browser.Mwebview;

public class MultiWebViewNMM : IOSNativeMicroModule
{
    static readonly Debugger Console = new("MultiWebViewNMM");

    public override string? ShortName { get; set; } = "MWebview";
    public MultiWebViewNMM() : base("mwebview.browser.dweb", "Multi Webview Renderer")
    {
    }

    public override List<MicroModuleCategory> Categories { get; init; } = new()
    {
        MicroModuleCategory.Service,
        MicroModuleCategory.Render_Service,
    };

    private static readonly Dictionary<Mmid, MultiWebViewController> s_controllerMap = new();
    public static MultiWebViewController? GetCurrentWebViewController(Mmid mmid) => s_controllerMap.GetValueOrDefault(mmid);

    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        /// nativeui 与 mwebview 是伴生关系
        await bootstrapContext.Dns.Open("nativeui.browser.dweb");

        /// 打开一个 webview，并将它以 窗口window 的标准进行展示
        HttpRouter.AddRoute(IpcMethod.Get, "/open", async (request, ipc) =>
        {
            var url = request.QueryStringRequired("url");
            var remoteMM = ipc?.AsRemoteInstance() ?? throw new Exception("mwebview.browser.dweb/open should be call by locale");

            ipc.OnClose += async (_) =>
            {
                Console.Log("/open", "listen ipc close destroy window");

                var controller = s_controllerMap.GetValueOrDefault(ipc!.Remote.Mmid);
                controller?.DestroyWebView();
            };

            var (viewItem, controller) = await OpenDwebViewAsync(remoteMM, url, ipc);

            return new ViewItemResponse(viewItem.webviewId, controller.Win.Id);
        });

        // 关闭指定 webview 窗口
        HttpRouter.AddRoute(IpcMethod.Get, "/close", async (request, ipc) =>
        {
            var searchParams = request.SafeUrl.SearchParams;
            var webviewId = searchParams.ForceGet("webview_id");
            var remoteMmid = ipc!.Remote.Mmid;

            return await _closeDwebViewAsync(remoteMmid, webviewId);
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/close/app", async (request, ipc) =>
        {
            var controller = s_controllerMap.GetValueOrDefault(ipc!.Remote.Mmid);

            if (controller is not null)
            {
                return controller.DestroyWebView();
            }

            return false;
        });

        // 界面没有关闭，用于重新唤醒
        HttpRouter.AddRoute(IpcMethod.Get, "/activate", async (request, ipc) =>
        {
            var remoteMmid = ipc!.Remote.Mmid;
            var controller = s_controllerMap.GetValueOrDefault(remoteMmid);

            if (controller is not null)
            {
                return false;
            }

            return new PureResponse(HttpStatusCode.OK);
        });
    }

    public record ViewItemResponse(string webviewId, UUID wid);

    private async Task<(MultiWebViewController.ViewItem, MultiWebViewController)> OpenDwebViewAsync(
        MicroModule remoteMM,
        string url,
        Ipc ipc)
    {
        var remoteMmid = remoteMM.Mmid;
        Console.Log("OPEN-WEBVIEW", "remote-mmid: {0} / url: {1}", remoteMmid, url);

        var controller = await MainThread.InvokeOnMainThreadAsync(() => s_controllerMap.GetValueOrPutAsync(remoteMmid, async () =>
        {
            var windowState = new WindowState(ipc.Remote.Mmid, Mmid, microModule: this);
            try
            {
                /// 挑选合适的图标作为应用的图标
                var iconResource = ipc.Remote.Icons?.Let(icons =>
                {
                    var comparableBuilder = ComparableWrapper<StrictImageResource>.Builder(imageResource =>
                    {
                        return new()
                        {
                            {
                                "purpose",
                                EnumComparable.EnumToComparable<ImageResourcePurposes>(imageResource.Purpose,
                                new List<ImageResourcePurposes> { ImageResourcePurposes.Any }).First()
                            },
                            {
                                "type",
                                EnumComparable.EnumToComparable(imageResource.Type,
                                new List<string> { "image/svg+xml", "image/png", "image/jpeg", "image/*" })
                            },
                            {
                                "area",
                                imageResource.Sizes.Last().Let(it => -it.Width * it.Height)
                            }
                        };
                    });

                    var imageResource = icons.Min();

                    if (imageResource is not null)
                    {
                        return comparableBuilder.Build(StrictImageResource.From(imageResource)).Value;
                    }

                    return null;
                });

                if (iconResource is not null)
                {
                    windowState.IconUrl = iconResource.Src;
                    windowState.IconMaskable = iconResource.Purpose.Contains(ImageResourcePurposes.Maskable);
                    windowState.IconMonochrome = iconResource.Purpose.Contains(ImageResourcePurposes.Monochrome);
                }
            }
            catch (Exception e)
            {
                Console.Error("windowstate", e.Message);
            }


            var win = await WindowAdapterManager.Instance.CreateWindow(windowState);
            return new MultiWebViewController(win, ipc, remoteMM, this).Also(controller =>
            {
                /// 窗口销毁的时候，释放这个Controller
                win.OnClose.OnListener += async (_, _) =>
                {
                    s_controllerMap.Remove(remoteMmid);
                };
            });
        }));


        return (await controller.OpenWebViewAsync(url), controller);
    }

    private async Task<bool> _closeDwebViewAsync(string remoteMmid, string webviewId)
    {
        var controller = s_controllerMap.GetValueOrDefault(remoteMmid);

        if (controller is not null)
        {
            //var vc = await RootViewController.WaitPromiseAsync();
            await MainThread.InvokeOnMainThreadAsync(async () =>
            {
                await controller.CloseWebViewAsync(webviewId);
                //await controller.DismissViewControllerAsync(true);
                //vc.PopViewController(true);
            });
        }

        return false;
    }
}

